src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,168 @@
+   :KeyV [:mode :visual]
+
+   ;; other
+   :KeyX        [:macro {:type :simple :value [[:attr nil] (:down-line -motions)]}]
+   ;; so these could be special or just, like... :tempo, and :octave
+   ;; I'm leaving them out for now
+   ;; :Period [:special [:tempo 1]]
+   ;; :ShiftPeriod [:special [:tempo -1]]
+   :Space [:special :play-pause]})
+
+(def edit-keymap
+  ;; hm, this isn't going to work quite like you want it to.
+  ;; this is fine for inputting notes...
+  {:KeyA [:attr :C]    :KeyW [:attr :C#]
+   :KeyS [:attr :D]    :KeyE [:attr :D#]
+   :KeyD [:attr :E]    :KeyR [:attr :F]
+   :KeyF [:attr :F#]   :KeyT [:attr :G]
+   :KeyG [:attr :G#]   :KeyY [:attr :A]
+   :KeyH [:attr :A#]   :KeyU [:attr :B]
+
+   ;; ... but when you get to the other attributes, what happens? They have
+   ;; values in [0x00, 0xff].
+   ;; You gotta either have another sort of translation layer where some of your
+   ;; note names become hex digits, or... something
+   ;; In fact, now that I think about it, this is why volume in the pre-alpha version
+   ;; only goes up to 9.
+   :Digit0 [:attr :Digit0]
+   :Digit1 [:attr :Digit1]
+   :Digit2 [:attr :Digit2]
+   :Digit3 [:attr :Digit3]
+   :Digit4 [:attr :Digit4]
+   :Digit5 [:attr :Digit5]
+   :Digit6 [:attr :Digit6]
+   :Digit7 [:attr :Digit7]
+   :Digit8 [:attr :Digit8]
+   :Digit9 [:attr :Digit9]
+
+   :KeyX      :off
+   :ShiftKeyX :stop
+
+   :Backspace [:macro {:type :simple :value [[:attr nil] (:up-line -motions)]}]
+   :Space     [:position (:down-line -motions)]})
+
+(def keymap
+  {:normal normal-keymap
+   :edit   edit-keymap})
+
+(def -movement-keys
+  #{:Space :ArrowDown :ArrowUp :ArrowLeft :ArrowRight :Tab :Backspace})
+
+(defn prevent-movement! [ev]
+  "Prevent movement keys (arrows, space, tab) from moving focus or scrolling."
+  (when (-movement-keys (keyword (.-code ev)))
+    (.preventDefault ev))
+  ev)
+
+(defn translate-keycode [ev]
+  "Takes keydown event and returns keycode as keyword; e.g., :KeyI, :Digit0.
+  If SHIFT is held, prepend Shift; e.g., :ShiftKeyI.
+  If CTRL is held, prepend Ctrl; e.g., :CtrlShiftKeyI."
+  (keyword (str (when (.-ctrlKey ev) "Ctrl")
+                (when (.-shiftKey ev) "Shift")
+                ;; .-code is nil if not a keypress
+                (.-code ev))))
+
+;; XXX don't know what to name these, but might not matter
+(defn position!
+  [params state])
+
+(defn frame!
+  [params state])
+
+(defn attr!
+  [params state])
+
+(defn macro!
+  [params state])
+
+(defn special!
+  [params state])
+
+(def handler-map
+  {:position position!
+   :frame frame!
+   :attr attr!
+   :special special!
+   :macro macro!})
+
+(defn handle-keypress!
+  "see top of file. idk, doc this if you want"
+  [ev state]
+  ;; when-let because it may not be a keypress. See `translate-keycode`
+  ;; the order of prevent-movement! and translate-keycode maybe can be changed?
+  ;; to shortcircuit a bit
+  (when-let [keycode (-> ev prevent-movement! translate-keycode)]
```

Comment:
I still find it kind of annoying that we handle `prevent-movement` in the same file we're handling "actual" user input, but it's fine for now

in the sense that "prevent movement" has nothing to do with notes, modes, commands, etc., it's just an artifact of the js environment

Reply:
Yeah, it would be nice to put it somewhere further up the chain. Gotta find a place.

Reply:
building off of my earlier comment about having an `actions` queue we dispatch off of, we could do something like

```
keycode -> (keycodes queue)

(keycodes queue) -> action-for-keypress -> (actions queue)
                 -> prevent-default-handler
```

i.e. add separate consumers of the keycodes queue for each separate "class" of actions we want to take off of it.

Then this module can just define stuff related to `actions` and the like, and we can put `keycode` and `action-for-keypress` and `prevent-default-handler` in another module.

Reply:
bump on this, I think we should definitely do a design like this. simplifies later features tremendously

Reply:
Ye, agreed.